Map remediation points to issues

ABaldwinHunter opened the pull request:
@codeclimate/review @mrb 

This change follows addition of remediation points to
[rubocop engine](https://github.com/codeclimate/codeclimate-rubocop/pull/22/files)

See [config/points.yml](https://github.com/codeclimate/codeclimate-brakeman/blob/abh-remediation-points/config/points.yml#L7) in this PR for a comment about how remediation
points translate to grades for constants in app.

Context: our goal is parity between security analysis on the legacy
and engines platform in terms of quality. However, on the [legacy platform](https://codeclimate.com/repos/560af5cfe30ba039cc000454/security),
brakeman security issues were reported separately from GPA, so this is a change.

Ex results from Brakeman engine:
<img width="1046" alt="screen shot 2015-09-30 at 4 19 38 pm" src="https://cloud.githubusercontent.com/assets/8718443/10206190/a1a06cce-6793-11e5-8317-409dc32879df.png">

Caveats:

Remediation cost hard to decide on: While many security issues
seem easy to fix - suggesting a low remediation cost -
they are typically more sensitive than other concerns.

Currently, our process requires engine issues to
have a line number. However, some brakemen issues do not.
For instance, 'protect_from_forgery' should be called in ApplicationController

For issues without begin and end lines, I have set defaults for now as 1.

TODO - add in a sustainable scraping mechanism similar to
rubocop https://github.com/codeclimate/codeclimate-rubocop/pull/26/files
that would grab all issue names and contents.

TODO - determine grade tuning for constants with security vulnerabilities.
Should one issue fail the constant?

TODO: Once desired behavior is determined, add tests

ABaldwinHunter commented on lib/cc/engine/brakeman.rb:
```diff
@@ -52,6 +53,28 @@ def content_body(link)
+
+        formula = points_hash[name]
+        # formula.try(:gsub!, '_', '')
+        # return unless formula.present?
```
I left this option for now. At present, an issue gets a number with no calculation.

ABaldwinHunter commented:
tests pass locally

```console
Ashleys-MacBook-Pro-2:codeclimate-brakeman ashleybaldwin-hunter$ be rake

# Running tests with run options --seed 11129:
....

Finished tests in 0.331449s, 12.0682 tests/s, 12.0682 assertions/s.

4 tests, 4 assertions, 0 failures, 0 errors, 0 skips
```

WIP [pull](https://github.com/codeclimate/codeclimate-brakeman/pull/5) to add circle testing

wfleming commented:
> Currently, our processes require engine issues to
> have a line number. However, some brakemen issues do not.
> For instance, 'protect_from_forgery' should be called in ApplicationController

This reminds me of an issue @mrb  had with foodcritic where it required a `README.md` to be present, and we crashed trying to read the non-existent file at some stage. Seems like the spec & our assumptions about issues are too restrictive.

ABaldwinHunter commented:
@wfleming nice comparison. I think I tend to agree

mrb commented:
@wfleming @ABaldwinHunter We can't currently handle issues that are assigned to a file/constant that doesn't exist. We can either creatively try to route them somwhere else, or ignore them.

wfleming commented:
@mrb I'm think that's a bug in our spec. Of course we can work around it now somehow or other, but it does seem like future work to keep in mind.

mrb commented:
@wfleming 100% agree.

mrb commented:
@wfleming Potential ideas include "repo level issues" and also just the ability to surface files even if they don't exist

ABaldwinHunter commented:
@codeclimate/review @mrb ready for re-review.

Updated spec - and added a `confidence` level detail to message output. (Probably not ideal solution long term, but gets the information across quickly.)

```CONSOLE
Ashleys-MacBook-Pro-2:railsgoat ashleybaldwin-hunter$ codeclimate analyze --dev -e brakeman
Starting analysis
Running brakeman: Done!

== app/controllers/api/v1/mobile_controller.rb (2 issues) ==
9: Unsafe reflection method constantize called with parameter value with confidence: Weak [brakeman]
16: Unsafe reflection method constantize called with parameter value with confidence: Weak [brakeman]

== app/controllers/application_controller.rb (1 issue) ==
1: 'protect_from_forgery' should be called in ApplicationController with confidence: Weak [brakeman]

== app/controllers/benefit_forms_controller.rb (2 issues) ==
10: Unsafe reflection method constantize called with parameter value with confidence: Weak [brakeman]
11: Parameter value used in file name with confidence: Weak [brakeman]

== app/controllers/dashboard_controller.rb (2 issues) ==
14: User controlled method execution with confidence: Weak [brakeman]
27: Render path contains parameter value with confidence: High [brakeman]

== app/controllers/password_resets_controller.rb (1 issue) ==
5: Marshal.load called with parameter value with confidence: Medium [brakeman]

== app/controllers/users_controller.rb (2 issues) ==
35: Possible SQL injection with confidence: Weak [brakeman]
58: Parameters should be whitelisted for mass assignment with confidence: Medium [brakeman]

== app/models/analytics.rb (1 issue) ==
2: Possible SQL injection with confidence: Medium [brakeman]

== app/models/benefits.rb (1 issue) ==
14: Possible command injection with confidence: Medium [brakeman]

== app/models/user.rb (1 issue) ==
17: Insufficient validation for 'email' using /.+@.+\..+/i. Use \A and \z as anchors with confidence: Weak [brakeman]

== app/views/layouts/application.html.erb (1 issue) ==
12: Unescaped cookie value with confidence: Weak [brakeman]

Analysis complete! Found 14 issues.
```

ABaldwinHunter commented on .ruby-version:
```diff
@@ -1 +1 @@
-2.2.3
+2.2.2
```
I switched to 2.2.2 for

1) consistency with our other engines 
2) ran into issues when I tried to install `rbx-2.2.3`

Not sure if there was a meaningful reason for going with 2.2.3 - welcome to input.

ABaldwinHunter commented on Gemfile.lock:
```diff
@@ -49,10 +54,12 @@ PLATFORMS
 DEPENDENCIES
   brakeman
+  dentaku
   minitest
```
This calculator tool isn't really needed for current functionality, but I thought it might be useful to include to pave the way for more complex remediation point calculations - e.g. `DentakuCalculator.new.store(SQL: 50000000)`. (There's already a movement afoot to use it in [rubocop](https://github.com/codeclimate/codeclimate-rubocop/pull/22/files#diff-8b7db4d5cc4b8f6dc8feb7030baa2478R4))

Might be YAGNI though.

wfleming replied:
I'm inclined to say this is YAGNI. The incorporation in rubocop is taking advantage of the fact that some of rubocop's metrics are quantifiable (e.g. complexity), so when that metric is higher, it's reasonable to expect it would take more effort to fix. But brakeman's only variable quantifiable attribute I see in the code is `confidence`, and that doesn't seem like an appropriate attribute to use to vary remediation points.

ABaldwinHunter replied:
@wfleming okay makes sense. Thanks for the feedback! I was wondering about categories of difficulty for brakeman fixes but don't have any in mind.

Removing dentaku for simpler model.

ABaldwinHunter replied:
Hm, that said, it also makes me question using a `yml` file to store points at all. If we expect issues to have equal remediation points, then we could set a constant number. `Brakeman.list_checks({})` gives all of the checks that can be performed dynamically, and also can take [optional](https://github.com/presidentbeef/brakeman/blob/master/lib/brakeman.rb#L233) additional checks that a [user has added](http://stackoverflow.com/questions/31799846/brakeman-rails-security-scanner-how-to-add-our-own-custom-check) - which might be cool to support.

ABaldwinHunter commented:
@codeclimate/review @mrb Ready for a final review. I made a rake task to populate the `points.yml` file with up to date brakeman checks, although it's not being used for anything yet. 

I also included Confidence level explanation in the readup.

Here are some shots:

In CLI:
<img width="881" alt="screen shot 2015-10-02 at 3 14 00 pm" src="https://cloud.githubusercontent.com/assets/8718443/10255538/0c4949c0-6919-11e5-958c-8d2affdf45e4.png">

On .com:
<img width="1051" alt="screen shot 2015-10-02 at 3 14 27 pm" src="https://cloud.githubusercontent.com/assets/8718443/10255560/32469e8e-6919-11e5-92d3-317350d4dc78.png">


results:
<img width="1238" alt="screen shot 2015-10-02 at 3 15 18 pm" src="https://cloud.githubusercontent.com/assets/8718443/10255546/1b69c826-6919-11e5-8813-603f15023202.png">

Readup:

<img width="1236" alt="screen shot 2015-10-02 at 3 15 31 pm" src="https://cloud.githubusercontent.com/assets/8718443/10255553/2499fd3a-6919-11e5-9cee-38b9f8b8b998.png">

GordonDiggs commented on lib/cc/engine/brakeman.rb:
```diff
@@ -4,6 +4,16 @@
 module CC
   module Engine
     class Brakeman
+      CONFIDENCE_LEVELS = [
```
:lipstick: Use `%w[]` here

GordonDiggs commented on lib/cc/engine/brakeman.rb:
```diff
@@ -4,6 +4,16 @@
+        "High"
+      ].freeze
+
+      CONFIDENCE_MESSAGE = "\#\# Brakeman Confidence that this vulnerability exists".freeze
```
Do these `#` need to be escaped?

ABaldwinHunter replied:
no :) thanks

GordonDiggs commented on lib/cc/engine/brakeman.rb:
```diff
@@ -30,26 +40,52 @@ def issue_json(warning)
           description: warning.message,
           categories: ["Security"],
-          remediation_points: 50_000,
+          confidence: warning.confidence,
```
Is this valid according to the spec?

wfleming replied:
*Technically* I don't think the spec says anything about including *extra* fields. So this is the best kind of correct. That said, I kind of doubt it will actually get pushed through correctly? Or properly used later? Might be something to insert into the `content`.

ABaldwinHunter replied:
Good questions. I'm checking.

The image built and ran on CLI and local harness analyses so it _works_ - checking into what happens with that key. (It's not being used atm, but I thought we might want that information later)

GordonDiggs commented on lib/cc/engine/brakeman.rb:
```diff
@@ -30,26 +40,52 @@ def issue_json(warning)
-        return File.read(path) if File.exist?(path)
-        "Read more: #{link}"
+
+        contents =
```
:lipstick: I'm sure our styleguide has something to say about this. Maybe indent it?

wfleming replied:
I believe our style guide discourages `var = if ...` overall. Might not catch this because of the line break?

ABaldwinHunter replied:
Updated to 
```
contents =
  if ...
```
I've seen some assignments [like](https://github.com/codeclimate/codeclimate/blob/master/lib/cc/analyzer/filesystem.rb#L41) that in our code, but maybe it's not desirable. I also considered a ternary operator (which I think we strongly eschew), and putting the variable assignment in each clause of conditional. 

Would something like the following be better?
```
if x
  content = func(x)
else
  content = y
end

confidence_report + content
```

wfleming replied:
Personally, I like

```
foo = if x
  :a
else
  :b
end
```

But I know I got snagged doing that in app at some point, so I don't think we like that style in-house. I think the newline make it worse, and harder to read. If `foo = if ...` is not permissible by our style, I think

```
if x
  foo = :a
else
  foo = :b
end
```

is the preferable style.

ABaldwinHunter replied:
:+1:

mrb commented:
@ABaldwinHunter Ok! I liked the confidence in the description FWIW /c @brynary

ABaldwinHunter commented:
Also experimented with adding brakeman confidence output in cli: https://github.com/codeclimate/codeclimate/pull/132
